src/credential-status-manager-base.ts

```diff
@@ -0,0 +1,572 @@
+    return Math.random().toString(36).substring(2,12).toUpperCase();
+  }
+
+  // embeds status into credential
+  async embedCredentialStatus({ credential, statusPurpose = 'revocation' }: EmbedCredentialStatusOptions): Promise<EmbedCredentialStatusResult> {
+    // ensure that credential has ID
+    if (!credential.id) {
+      // Note: This assumes that uuid will never generate an ID that
+      // conflicts with an ID that has already been tracked in the log
+      credential.id = uuid();
+    }
+
+    // find latest relevant log entry for credential with given ID
+    const logData: CredentialStatusLogData = await this.readLogData();
+    logData.reverse();
+    const logEntry = logData.find((entry) => {
+      return entry.credentialId === credential.id;
+    });
+
+    // do not allocate new status list entry if ID is already being tracked
+    if (logEntry) {
+      // retrieve relevant log data
+      const { statusListId: logStatusListId, statusListIndex } = logEntry;
+
+      // attach credential status
+      const statusUrl = this.getCredentialStatusUrl();
+      const statusListCredential = `${statusUrl}/${logStatusListId}`;
+      const statusListId = `${statusListCredential}#${statusListIndex}`;
+      const credentialStatus = {
+        id: statusListId,
+        type: CREDENTIAL_STATUS_TYPE,
+        statusPurpose,
+        statusListIndex,
+        statusListCredential
+      };
+
+      return {
+        credential: {
+          ...credential,
+          credentialStatus,
+          '@context': [...credential['@context'], CONTEXT_URL_V1]
+        }
+      };
+    }
+
+    // retrieve status config data
+    const configData = await this.readConfigData();
+    let { credentialsIssued, latestList } = configData;
+
+    // allocate new status list entry if ID is not yet being tracked
+    let newList;
+    if (credentialsIssued >= CREDENTIAL_STATUS_LIST_SIZE) {
+      latestList = this.generateStatusListId();
+      newList = latestList;
+      credentialsIssued = 0;
+    }
+    credentialsIssued++;
+
+    // update status config data
+    configData.credentialsIssued = credentialsIssued;
+    configData.latestList = latestList;
+    await this.updateConfigData(configData);
+
+    // attach credential status
+    const statusUrl = this.getCredentialStatusUrl();
+    const statusListCredential = `${statusUrl}/${latestList}`;
+    const statusListIndex = credentialsIssued;
+    const statusListId = `${statusListCredential}#${statusListIndex}`;
+    const credentialStatus = {
+      id: statusListId,
+      type: CREDENTIAL_STATUS_TYPE,
+      statusPurpose,
+      statusListIndex,
+      statusListCredential
+    };
+
+    return {
+      credential: {
+        ...credential,
+        credentialStatus,
+        '@context': [...credential['@context'], CONTEXT_URL_V1]
+      },
+      newList
+    };
+  }
+
+  // allocates status for credential
+  async allocateStatus(credential: VerifiableCredential): Promise<VerifiableCredential> {
```

Comment:
This function (and the previous one) doesn't feel quite right (in terms of length and some duplication of logic), might be a good candidate for future refactoring.